r1cs/gadgets/std/src/fields/nonnative/params.rs

```diff
@@ -1,169 +1,212 @@
-        let num_adds_prod: u128 = (num_limbs as u128*num_adds_plus_one*num_adds_plus_one)*(pseudo_mersenne_const as u128*2u128.pow(h as u32)+1)*2u128.pow(bits_per_limb as u32) - 1;
-        ceil_log2(num_adds_prod+3)+1
-    } else {
-        // otherwise, we compute the log without explicitly computing num_adds_prod:
-        // given num_adds_unshifted = num_limbs*(num_adds+1)*(num_adds+1), we need
-        // to compute 1+log(num_adds_unshifted*(c*2^h+1)*2^bits_per_limb + 2). Since num_adds_unshifted
-        // must be shifted by bits_per_limb, we can neglect the addition with 2 and compute
-        // log(num_adds_unshifted*2^bits_per_limb + 2) = log(num_adds_unshifted*(c*2^h+1))+bits_per_limb.
-        // To compute log(num_adds_unshifted), we consider that
-        // num_adds_unshifted*(c*2^h+1) = num_adds_unshifted*c*2^h + num_adds_unshifted.
-        // Since num_adds_unshifted < 2^31 and c < 2^64, then if h < 32, num_adds_unshifted*(c*2^h+1)
-        // can be computed without overflow. Otherwise, if h >= 32, then
-        // log(num_adds_unshifted*(c*2^h+1)) = log(num_adds_unshifted*c*2^h) = log(num_adds_unshifted*c)+h
-        let log_num_adds_unshifted = if h < 32 {
-            let num_adds_unshifted: u128 = num_limbs as u128 * num_adds_plus_one * num_adds_plus_one * (pseudo_mersenne_const as u128*2u128.pow(h as u32) + 1);
-            128 - num_adds_unshifted.leading_zeros() as usize
-            // NOTE: no need to apply correction to the result if num_adds_unshifted is a power of two,
-            // since num_adds_unshifted*2^bits_per_limb + 2 is never a power of two
+    let capacity = base_field_prime_length - 1;
+
+    let mut bits_per_limb = 1usize;
+
+    while bits_per_limb <= target_field_prime_length - 1 {
+        // we compute the number of constraints for a `single mul_without_prereduce()`
+        // and a subsequent `reduce()`
+        let num_limbs = (target_field_prime_length + bits_per_limb - 1) / bits_per_limb;
+        let h = bits_per_limb * num_limbs - target_field_prime_length;
+        // cost of mul
+        let mut constraints = num_limbs * num_limbs;
+        // compute surfeit_prod in a compatible way with const function:
+        // surfeit_prod = 1 + log(num_limbs*(num_adds+1)*(num_adds+1)*(c*2^h+1)*2^bits_per_limb + 2)
+        // surfeit_prod may overflow u128 if bits_per_limb is high
+        let surfeit_prod = if bits_per_limb + h < 32 {
+            // in this case we can compute num_adds_prod without overflowing, as num_limbs*(num_adds+1)*(num_adds+1)*(c+1) < 2^96 with our choice of surfeit
+            let num_adds_prod: u128 = (num_limbs as u128 * num_adds_plus_one * num_adds_plus_one)
+                * (pseudo_mersenne_const as u128 * 2u128.pow(h as u32) + 1)
+                * 2u128.pow(bits_per_limb as u32)
+                - 1;
+            ceil_log2(num_adds_prod + 3) + 1
```

Comment:
Can't we just throughout use the generic code from the `else` clause?

Reply:
I think it will not work with really low values of bits_per_limb (e.g., bits_per_limb=1), which are still evaluated by the function. Indeed the generic code in the else clause to compute the logarithm relies on the fact that since `num_adds_unshifted` is shifted by 'bits_per_limb`, then the addition of 2 has no effect on the logarithm, but I think this is true only if `bits_per_limb > 1`. So we would still need to have a specific branch to handle the case when `bits_per_limb = 1`.

Reply:
Ok, let us keep with it.